File: adapters/ios-keyboard/swift/KhmerIMEKeyboard/Input/SmartModePreference.swift
import Foundation

// SmartModePreference
// ===================
// Persists the user's Standard/Smart choice in a shared App Group UserDefaults suite, so the
// host-app Settings toggle (writer) and the keyboard extension (reader) see the same value.
// Standard (the default) is lexicon + fuzzy only; Smart enables the injected span-proposal
// provider — inert without a registered provider, so in the OSS build the toggle has no visible
// effect. Provider-agnostic: names no model. Mirrors the Android SmartModePreference.

struct SmartModePreference {

    // The App Group shared between the app and keyboard extension (see project.yml entitlements).
    static let suiteName = "group.com.khmerime"
    private static let key = "smart_mode"

    private let defaults: UserDefaults

    // `defaults` is injectable for tests; production uses the shared App Group suite.
    init(defaults: UserDefaults = UserDefaults(suiteName: SmartModePreference.suiteName) ?? .standard) {
        self.defaults = defaults
    }

    var isEnabled: Bool {
        defaults.bool(forKey: SmartModePreference.key)
    }

    func setEnabled(_ enabled: Bool) {
        defaults.set(enabled, forKey: SmartModePreference.key)
    }
}

File: adapters/android-ime/app/src/main/java/com/khmerime/input/SmartModePreference.kt
package com.khmerime.input

import android.content.Context

// SmartModePreference
// ===================
// Persists the user's Standard/Smart choice. Standard = lexicon + fuzzy only (the default). Smart
// enables the injected span-proposal provider inside the decoder — inert (a no-op) unless a provider
// is registered, so in the OSS build this toggle has no visible effect and the engine stays Standard.
// Mirrors the iOS SmartModePreference. Provider-agnostic: names no model.

object SmartModePreference {

    private const val PREFS_NAME = "khmerime_settings"
    private const val KEY_SMART_MODE = "smart_mode"

    fun isEnabled(context: Context): Boolean =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getBoolean(KEY_SMART_MODE, false)

    fun setEnabled(context: Context, enabled: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_SMART_MODE, enabled)
            .apply()
    }
}
